realm/realm-library/src/androidTestObjectServer/kotlin/io/realm/RealmFunctionsTests.kt

```diff
@@ -0,0 +1,150 @@
+                    //  io.realm.exceptions.RealmError: Unrecoverable error. current state '$1' is not of expected state '$2' in /Users/claus.rorbech/proj/realm-java/realm/realm-library/src/main/cpp/io_realm_RealmFunctions.cpp line 32
+                    //val listValues = listOf<Any>(true, i32, i64)
+                    val listValues = listOf<Any>(i32, i64)
+                    assertTypedEcho(listValues, List::class.java)
+                }
+                BsonType.BINARY -> {
+                    val value = byteArrayOf(1, 2, 3)
+                    val actual = functions.invoke(value, ByteArray::class.java)
+                    assertEquals(value.toList(), actual.toList())
+                    // FIXME Does not seem to preserve type
+                    // assertTypedEcho(BsonBinary(UUID.randomUUID()), BsonBinary::class.java)
+                    assertTypedEcho(BsonBinary(byteArrayOf(1,2,3)), BsonBinary::class.java)
+                }
+                BsonType.OBJECT_ID -> {
+                    assertTypedEcho(ObjectId(), ObjectId::class.java)
+                    assertTypedEcho(BsonObjectId(ObjectId()), BsonObjectId::class.java)
+                }
+                BsonType.BOOLEAN -> {
+                    val value: Boolean = true
+                    val actual: java.lang.Boolean = functions.invoke(value, java.lang.Boolean::class.java)
+                    assertEquals(value, actual.booleanValue())
+                    assertTypedEcho(BsonBoolean(true), BsonBoolean::class.java)
+                }
+                BsonType.INT32 -> {
+                    assertTypedEcho(java.lang.Integer(32), Integer::class.java)
+                    assertTypedEcho(BsonInt32(32), BsonInt32::class.java)
+                }
+                BsonType.INT64 -> {
+                    assertTypedEcho(java.lang.Long(32L), java.lang.Long::class.java)
+                    assertTypedEcho(BsonInt64(32), BsonInt64::class.java)
+                }
+                BsonType.DECIMAL128 -> {
+                    assertTypedEcho(Decimal128(32L), Decimal128::class.java)
+                    assertTypedEcho(BsonDecimal128(Decimal128(32L)), BsonDecimal128::class.java)
+                }
+                // TODO
+                BsonType.DOCUMENT,
+                BsonType.UNDEFINED,
+                BsonType.DATE_TIME,
+                BsonType.NULL,
+                BsonType.REGULAR_EXPRESSION,
+                BsonType.SYMBOL,
+                BsonType.DB_POINTER,
+                BsonType.JAVASCRIPT,
+                BsonType.JAVASCRIPT_WITH_SCOPE,
+                BsonType.TIMESTAMP,
+                BsonType.END_OF_DOCUMENT,
+                BsonType.MIN_KEY,
+                BsonType.MAX_KEY -> {
+                    // No conversion is implemented for these types yet
+                }
+            }
+        }
+    }
+
+    private fun <T: Any> assertTypedEcho(value: T, returnClass: Class<T>) : T {
```

Comment:
minor: Not sure I understand the Echo suffix. Am I missing some commonly known convention here?

Reply:
No. Just because the underlying JNI implementation is just echo'ing (returning the arguments directly) as it is not yet hitting the action Stitch functions .